### DISCUSSION TITLE
Changing the style of the selected custom nodes

### DISCUSSION THREAD

**Discussion by SciManJD** [bcakmakoglu/vue-flow#903] [Q&A] [answered]
I created several custom nodes and when I click to them, they are selected, but this is not visible. When I click to a standard node, its style is changing. I want to do the same for my new custom nodes. Probably it is quite easy, but right now I cannot imaging how to do this. Can I get some help, please?

**Comment by bcakmakoglu** [comment, marked as answer]
[Here's an example of styling a selected custom node.](https://codesandbox.io/p/sandbox/mystifying-lewin-u9xjw9)

You could also use css to style the wrapper of your custom node

```css
.vue-flow__node-custom.selected {
	/* your styles */
}
```

**Reply by SciManJD** [reply]
Sorry, but I still cannot get the complete picture. Who controls this property "customNodeProps.selected"? I already use a property "data" that sends the label and some mode data to the custom component. How to combine this with the property "customNodeProps.selected"?

**Reply by bcakmakoglu** [reply]
[Custom node props](https://vueflow.dev/typedocs/interfaces/NodeProps.html) are passed by Vue Flow to your custom node components.

If you use a slot you can use these from the slot props
```vue
<script setup>
import { VueFlow } from '@vue-flow/core'

const elements = ref([
	{
		id: '1',
		type: 'custom',
		label: 'Node 1',
		position: { x: 0, y: 0 }
		data: { foo: true }
	},
	{
		id: '2',
		type: 'custom',
		label: 'Node 2',
		position: { x: 250, y: 0 }
		data: { foo: true }
	},
	{
		id: 'e1-2',
		source: '1',
		target: '2',
	},
])
</script>

<template>
	<VueFlow v-model="elements">
		<template #node-custom="customNodeProps">
			<!-- pass all props at once -->
			<CustomNode v-bind="customNodeProps" />
		
			<!-- only pass specific props -->
			<CustomNode :selected="customNodeProps.selected" :label="customNodeProps.label" />
		</template>
	</VueFlow>
</template>
```

If you're using a node types object, all props are automatically passed to your custom node.

You can access these props from your custom node like this

```vue
<script setup>
// use the `label` and `selected` props, which are passed as either slotprops (`customNodeProps` in the code above) or just passed to your component directly (for example using `v-bind` or a node types object)
defineProps(['label', 'selected'])
</script>

<template>
	<div :class="{ selected }">{{ label }}</div>
</template>
```

**Reply by bcakmakoglu** [reply]
As to the `selected` flag - when a node is clicked, dragged, clicked while multi-selecting or included in the selection box (shift + click & drag) it will get the `selected` flag.
This flag will toggle a class on the wrapper element, i.e. `.vue-flow__node-${type}` & `.selected`.
It will also be passed down to custom node components as a prop.

**Reply by SciManJD** [reply]
I use the slot to add my custom component. I also pass all props like this:
```
<template>
	<VueFlow v-model="elements">
		<template #node-custom="data">
			<!-- pass all props at once -->
			<CustomNode v-bind="data" />
		</template>
	</VueFlow>
</template>
```
So I expect that this flag 'selected' will be passed to the custom component automatically.
Then in my component I defined property 'data' and I can access it without problems. Now I added also a property 'selected' and a watcher to be sure that I really get the flag: 

```
    export default {
        ...,
        props: {
            data: {
                type: Object,
                required: true
            },
            selected: Boolean
        },
        watch: {
            selected: function(state) {
                console.log(state)
            }
        },
   }
```

Ans I expect that when I click on the custom node, I will see a print, but it never occurs. Sorry, but I still don't get it...

**Reply by SciManJD** [reply]
There is a mistake in my code (I suppose)... Probably the 'selected' property is an element of 'data'. I tried this too, but still cannot catch the 'selected' property.

**Reply by SciManJD** [reply]
My watcher was also wrong. I fixed it:

```
        watch: {
            // Watch the property (data) given from the parent component
            data: {
                deep: true,
                immediate: true,
                handler(newData) {
                    console.log(JSON.stringify(newData))
                }
            }
        },

```

Now I get prints when the 'data' is changed, but the 'selected' is not inside.

**Reply by bcakmakoglu** [reply]
Again: The slot props are of the type `NodeProps` which I already linked above, they are not the data object.
The data object is *part* of the slot props.

So your assumption that `#node-custom="data"` is just wrong, because the reality is it is a property of the props, i.e. `#node-custom="{ data }"` - same as the selected flag.

**Reply by bcakmakoglu** [reply]
[Please read the type of `NodeProps` properly, so you actually understand what the props actually include.](https://vueflow.dev/typedocs/interfaces/NodeProps.html)

**Reply by SciManJD** [reply]
Now it works. I read the section for NodeProps several type (it is not too big :)), but I couldn't understand it. But if you say that my 'data' is a part of the NodeProps, it became clear. Thanks a lot for your patience. I told you that I'm still learning... sorry...

**Reply by bcakmakoglu** [reply]
No worries I just can't write too in-depth from my phone 🤣

**Reply by SciManJD** [reply]
Short additional question. All works fine when I want to get the 'selected' property inside the custom node. But now I tried to change the wrapper class. So according to your notes, I added a new class in the file that VueFlow component is included:

```
<template>
    <vue-flow> 
        <template #node-custom="customNodeProps">
            <vue-flow-custom v-bind="customNodeProps" @input="handleInputChange"/>
        </template>
    </vue-flow>
</template>

<style scoped>
   .vue-flow__node-${custom}.selected {
        outline: none;
        border: 1px solid var(--vf-node-color);
        box-shadow: 0 0 0 .5px var(--vf-box-shadow);
   }
</style>
```

But I can't compile this... Probably this is not the syntax that you meant, but this is how I understood it... Help, please...

**Reply by bcakmakoglu** [reply]
the `${var}` syntax is a js syntax not css (it was just to used to demonstrate what the class name should look like)

So the class name should look like this:
```css
.vue-flow__node-custom.selected { /* ... */ }
```

**Reply by SciManJD** [reply]
This was the first thing that I tried, but it didn't work. I know that ${var} is not a css syntax, but I doubt in my knowledge, that's why I tried it. If I use .vue-flow__node-custom.selected, I can build, but there is no change when I click on the node. What else could I miss?

**Reply by SciManJD** [reply]
I found the problem. My class was in scoped-section of the styles. But it seems that it must be global. All works fine...
